### PR TITLE
[cmake] add lapack,libmct,mklml third_party cache

### PR DESCRIPTION
**Pull request by gouzil**
### PR types

Others
### PR changes

Others
### Description

添加`lapack`,`libmct`,`mklml`库编译缓存

相关链接:
 * #54305
 * [支持Paddle第三方库离线编译](https://github.com/PaddlePaddle/community/blob/master/pfcc/call-for-contributions/code_style/support_offline_compile_of_third_party.md)

### CODE REVIEW

File: cmake/external/lapack.cmake [hunk 1]
```diff
@@ -15,18 +15,19 @@
 include(ExternalProject)
 
 set(LAPACK_PREFIX_DIR ${THIRD_PARTY_PATH}/lapack)
-set(LAPACK_SOURCE_DIR ${THIRD_PARTY_PATH}/lapack/src/extern_lapack)
+set(LAPACK_DOWNLOAD_DIR
+    ${PADDLE_SOURCE_DIR}/third_party/lapack/${CMAKE_SYSTEM_NAME})
```
Comment (gouzil) [inline cmake/external/lapack.cmake:19]: 这里是为了防止，有用户会在docker和本机同时编译所以在这里以系统分文件夹, 比如在macos同时使用docker和本机

目录结构
```txt
├── ...
└── lapack/
    ├── .../
    |── Darwin/
    |     ├── ...
    |     └──xxx.gz
    └── Linux/
         ├── ...
         └──xxx.gz
```

File: cmake/external/lapack.cmake [hunk 2]
```diff
@@ -62,18 +63,53 @@ else()
...
+  NAMES ${LAPACK_VER}
+  PATHS ${LAPACK_DOWNLOAD_DIR}
+  NO_DEFAULT_PATH)
+
+# Download and check lapack.
+if(LOCAL_LAPACK_LIB_ZIP)
+  file(MD5 ${LAPACK_DOWNLOAD_DIR}/${LAPACK_VER} LAPACK_MD5)
+  if(NOT LAPACK_MD5 EQUAL LAPACK_URL_MD5)
+    download_lapack()
+  endif()
```
Comment (gouzil) [inline cmake/external/lapack.cmake]: 这里是为了校验是否有更新, 如果更新就会重新下载, 这里下载的时候可以考虑是否需要删除之前的文件(同名会覆盖)

File: cmake/external/libmct.cmake [hunk 3]
```diff
@@ -39,22 +43,27 @@ set(CMAKE_INSTALL_RPATH "${CMAKE_INSTALL_RPATH}" "${LIBMCT_ROOT}/lib")
...
   PREFIX ${LIBMCT_PREFIX_DIR}
   DOWNLOAD_DIR ${LIBMCT_DOWNLOAD_DIR}
-  DOWNLOAD_COMMAND
-    wget --no-check-certificate ${LIBMCT_URL} -c -q -O ${LIBMCT_NAME}.tar.gz &&
-    tar --no-same-owner -zxvf ${LIBMCT_NAME}.tar.gz
-  DOWNLOAD_NO_PROGRESS 1
+  SOURCE_DIR ${LIBMCT_INSTALL_DIR}
   UPDATE_COMMAND ""
+  COMMAND ${CMAKE_COMMAND} -E copy ${LIBMCT_DOWNLOAD_DIR}/CMakeLists.txt
+          ${LIBMCT_INSTALL_DIR}
```
Comment (gouzil) [inline cmake/external/libmct.cmake:66]: 这里用拷贝是因为当`make -j`时会清空`build/third_parth/install/libmct`文件夹

File: cmake/external/lapack.cmake (hunk 1 above)
Comment (risemeup1) [reply cmake/external/lapack.cmake:19]: It's a great idea, and we didn't think about that before

**Review by risemeup1 — APPROVED** [review]
LGTM，there are some problems with the Coverage CI, but they are not caused by your PR.I helped you rerun , after CI passed to help you merge

**Comment by gouzil** [comment]
> LGTM，there are some problems with the Coverage CI, but they are not caused by your PR.I helped you rerun , after CI passed to help you merge

If cross compilation is supported in the future, Then it is necessary to continue distinguishing between CPU architectures

**Review by risemeup1 — DISMISSED** [review]

File: cmake/external/lapack.cmake [hunk 4]
```diff
@@ -62,18 +63,53 @@ else()
...
+  DOWNLOAD_DIR ${LAPACK_DOWNLOAD_DIR}
+  SOURCE_DIR ${LAPACK_LIB_DIR}
   PREFIX ${LAPACK_PREFIX_DIR}
   DOWNLOAD_NO_PROGRESS 1
   PATCH_COMMAND ""
   UPDATE_COMMAND ""
   CONFIGURE_COMMAND ""
   BUILD_COMMAND ""
-  INSTALL_COMMAND ${CMAKE_COMMAND} -E copy_directory ${LAPACK_SOURCE_DIR}
-                  ${LAPACK_LIB_DIR}
```
Comment (risemeup1) [inline cmake/external/lapack.cmake:77]: 这个直接删除了没有影响吗？

File: cmake/external/lapack.cmake [hunk 5]
```diff
@@ -49,10 +50,10 @@ elseif(WIN32)
   set(LAPACK_LIB "${LAPACK_LIB_DIR}/liblapack.dll")
 else()
   set(LAPACK_VER
```
Comment (zhwesky2010) [inline cmake/external/lapack.cmake]: 这个如果把版本改成带后缀的文件名，那最好就换个变量名，比如LAPACK_FILE之类

Comment (gouzil) [reply cmake/external/lapack.cmake]: Done

File: cmake/external/lapack.cmake (hunk 4 above)
Comment (gouzil) [reply cmake/external/lapack.cmake:77]: 这里没问题，因为指定了`SOURCE_DIR`直接就解压到原先的`LAPACK_LIB_DIR`里面了

**Review by risemeup1 — APPROVED** [review]

**Pull request merged**